File: Ports/jdk/jdk-jdk-18-10/src/java.base/share/classes/jdk/internal/icu/util/VersionInfo.java
/*
 * Copyright (c) 2005, 2020, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
/*
 *******************************************************************************
 * (C) Copyright IBM Corp. and others, 1996-2009 - All Rights Reserved         *
 *                                                                             *
 * The original version of this source code and documentation is copyrighted   *
 * and owned by IBM, These materials are provided under terms of a License     *
 * Agreement between IBM and Sun. This technology is protected by multiple     *
 * US and International patents. This notice and attribution to IBM may not    *
 * to removed.                                                                 *
 *******************************************************************************
 */

package jdk.internal.icu.util;

import java.util.HashMap;

/**
 * Class to store version numbers of the form major.minor.milli.micro.
 * @author synwee
 * @stable ICU 2.6
 */
public final class VersionInfo
{
    // public data members -------------------------------------------------

    /**
     * Data version string for ICU's internal data.
     * Used for appending to data path (e.g. icudt43b)
     * @internal
     * @deprecated This API is ICU internal only.
     */
    @Deprecated
    public static final String ICU_DATA_VERSION_PATH = "67b";

    // public methods ------------------------------------------------------

    /**
     * Returns an instance of VersionInfo with the argument version.
     * @param version version String in the format of "major.minor.milli.micro"
     *                or "major.minor.milli" or "major.minor" or "major",
     *                where major, minor, milli, micro are non-negative numbers
     *                {@literal <=} 255. If the trailing version numbers are
     *                not specified they are taken as 0s. E.g. Version "3.1" is
     *                equivalent to "3.1.0.0".
     * @return an instance of VersionInfo with the argument version.
     * @exception throws an IllegalArgumentException when the argument version
     *                is not in the right format
     * @stable ICU 2.6
     */
    public static VersionInfo getInstance(String version)
    {
        int length  = version.length();
        int array[] = {0, 0, 0, 0};
        int count   = 0;
        int index   = 0;

        while (count < 4 && index < length) {
            char c = version.charAt(index);
            if (c == '.') {
                count ++;
            }
            else {
                c -= '0';
                if (c < 0 || c > 9) {
                    throw new IllegalArgumentException(INVALID_VERSION_NUMBER_);
                }
                array[count] *= 10;
                array[count] += c;
            }
            index ++;
        }
        if (index != length) {
            throw new IllegalArgumentException(
                                               "Invalid version number: String '" + version + "' exceeds version format");
        }
        for (int i = 0; i < 4; i ++) {
            if (array[i] < 0 || array[i] > 255) {
                throw new IllegalArgumentException(INVALID_VERSION_NUMBER_);
            }
        }

        return getInstance(array[0], array[1], array[2], array[3]);
    }

    /**
     * Returns an instance of VersionInfo with the argument version.
     * @param major major version, non-negative number {@literal <=} 255.
     * @param minor minor version, non-negative number {@literal <=} 255.
     * @param milli milli version, non-negative number {@literal <=} 255.
     * @param micro micro version, non-negative number {@literal <=} 255.
     * @exception throws an IllegalArgumentException when either arguments are
     *                                     negative or {@literal >} 255
     * @stable ICU 2.6
     */
    public static VersionInfo getInstance(int major, int minor, int milli,
                                          int micro)
    {
        // checks if it is in the hashmap
        // else
        if (major < 0 || major > 255 || minor < 0 || minor > 255 ||
            milli < 0 || milli > 255 || micro < 0 || micro > 255) {
            throw new IllegalArgumentException(INVALID_VERSION_NUMBER_);
        }
        int     version = getInt(major, minor, milli, micro);
        Integer key     = Integer.valueOf(version);
        Object  result  = MAP_.get(key);
        if (result == null) {
            result = new VersionInfo(version);
            MAP_.put(key, result);
        }
        return (VersionInfo)result;
    }

    /**
     * Compares other with this VersionInfo.
     * @param other VersionInfo to be compared
     * @return 0 if the argument is a VersionInfo object that has version
     *           information equal to this object.
     *           Less than 0 if the argument is a VersionInfo object that has
     *           version information greater than this object.
     *           Greater than 0 if the argument is a VersionInfo object that
     *           has version information less than this object.
     * @stable ICU 2.6
     */
    public int compareTo(VersionInfo other)
    {
        return m_version_ - other.m_version_;
    }

    // private data members ----------------------------------------------

    /**
     * Version number stored as a byte for each of the major, minor, milli and
     * micro numbers in the 32 bit int.
     * Most significant for the major and the least significant contains the
     * micro numbers.
     */
    private int m_version_;
    /**
     * Map of singletons
     */
    private static final HashMap<Integer, Object> MAP_ = new HashMap<>();
    /**
     * Error statement string
     */
    private static final String INVALID_VERSION_NUMBER_ =
        "Invalid version number: Version number may be negative or greater than 255";

    // private constructor -----------------------------------------------

    /**
     * Constructor with int
     * @param compactversion a 32 bit int with each byte representing a number
     */
    private VersionInfo(int compactversion)
    {
        m_version_ = compactversion;
    }

    /**
     * Gets the int from the version numbers
     * @param major non-negative version number
     * @param minor non-negativeversion number
     * @param milli non-negativeversion number
     * @param micro non-negativeversion number
     */
    private static int getInt(int major, int minor, int milli, int micro)
    {
        return (major << 24) | (minor << 16) | (milli << 8) | micro;
    }
}

File: Ports/jdk/jdk-jdk-18-10/src/jdk.jpackage/windows/native/libjpackage/VersionInfo.cpp
/*
 * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

#include <sstream>
#include <iomanip>
#include "VersionInfo.h"
#include "ResourceEditor.h"
#include "ErrorHandling.h"
#include "Toolbox.h"


VersionInfo::VersionInfo() {

}

namespace {

class FixedFileVersion {
public:
    FixedFileVersion(const std::wstring& value) {
        if (4 != swscanf_s(value.c_str(), L"%d.%d.%d.%d", components + 0,
                components + 1, components + 2, components + 3)) {
            JP_THROW(tstrings::any()
                    << "Malformed file version value: ["
                    << value
                    << "]");
            forEach(components, [&value](int component) -> void {
                if (USHRT_MAX < component) {
                    JP_THROW(tstrings::any()
                        << "Invalid file version value: ["
                        << value
                        << "]");
                }
            });
        }
    }

    void apply(DWORD& ms, DWORD& ls) const {
        ms = MAKELONG(components[1], components[0]);
        ls = MAKELONG(components[3], components[2]);
    }

private:
    int components[4];
};


std::ostream& writeWORD(std::ostream& cout, size_t v) {
    if (USHRT_MAX < v) {
        JP_THROW("Invalid WORD value");
    }
    return cout.write(reinterpret_cast<const char*>(&v), sizeof(WORD));
}


std::ostream& writeDWORD(std::ostream& cout, size_t v) {
    if (UINT_MAX < v) {
        JP_THROW("Invalid DWORD value");
    }

    return cout.write(reinterpret_cast<const char*>(&v), sizeof(DWORD));
}


std::ostream& write(std::ostream& cout, const VS_FIXEDFILEINFO& v) {
    return cout.write(reinterpret_cast<const char*>(&v), sizeof(v));
}

std::ostream& write(std::ostream& cout, const std::wstring& s) {
    return cout.write(reinterpret_cast<const char*>(s.c_str()),
            (s.size() + 1 /* trailing 0 */) * sizeof(wchar_t));
}

void add32bitPadding(std::ostream& cout) {
    enum { WordAlign = 2 };
    const std::streampos pos = cout.tellp();
    if (pos % 2) {
        JP_THROW("Invalid data written in the stream");
    }
    const int padding = WordAlign - (pos / 2) % WordAlign;
    if (WordAlign != padding) {
        for (int i = 0; i < padding; ++i) {
            writeWORD(cout, 0);
        }
    }
}


class StreamSize {
public:
    StreamSize(std::ostream& out): stream(out), anchor(out.tellp()) {
        writeWORD(stream, 0); // placeholder
    }

    ~StreamSize() {
        JP_TRY;

        const std::streampos curPos = stream.tellp();
        const std::streampos size = curPos - anchor;
        stream.seekp(anchor);
        if (size < 0) {
            JP_THROW("Invalid negative size value");
        }
        writeWORD(stream, (size_t) size);
        stream.seekp(curPos);

        JP_CATCH_ALL;
    }

private:
    std::ostream& stream;
    std::streampos anchor;
};

} // namespace

VersionInfo& VersionInfo::setProperty(
        const std::wstring& id, const std::wstring& value) {
    props[id] = value;

    if (id == L"FIXEDFILEINFO_FileVersion") {
        // Validate input
        const ::FixedFileVersion validate(value);
    }
    return *this;
}


const VersionInfo& VersionInfo::apply(
        const ResourceEditor::FileLock& fileLock) const {
    if (props.find(L"FIXEDFILEINFO_FileVersion") == props.end()) {
        JP_THROW("Missing mandatory FILEVERSION property");
    }

    std::stringstream buf(
            std::stringstream::in | std::stringstream::out | std::stringstream::binary);
    buf.exceptions(std::ios::failbit | std::ios::badbit);

    fillBuffer(buf);

    buf.seekg(0);

    ResourceEditor()
        .id(MAKEINTRESOURCE(VS_VERSION_INFO))
        .type(RT_VERSION)
        .apply(fileLock, buf);
    return *this;
}


void VersionInfo::fillBuffer(std::ostream& buf) const {
    // Fill VS_VERSIONINFO pseudo structure
    StreamSize versionInfoLength(buf); // wLength
    writeWORD(buf, sizeof(VS_FIXEDFILEINFO)); // wValueLength
    writeWORD(buf, 0); // wType
    write(buf, L"VS_VERSION_INFO"); // szKey
    add32bitPadding(buf); // Padding1
    write(buf, createFIXEDFILEINFO()); // Value
    add32bitPadding(buf); // Padding2

    const DWORD neutralLangId = (0x04b0 | MAKELANGID(LANG_NEUTRAL, SUBLANG_NEUTRAL) << 16);
    const DWORD engLangId = (0x04b0 | MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US) << 16);

    do {
        // Fill StringFileInfo pseudo structure
        StreamSize stringFileInfoLength(buf); // wLength
        writeWORD(buf, 0); // wValueLength
        writeWORD(buf, 1); // wType
        write(buf, L"StringFileInfo"); // szKey
        add32bitPadding(buf); // Padding

        // Fill StringTable pseudo structure
        StreamSize stringTableLength(buf); // wLength
        writeWORD(buf, 0); // wValueLength
        writeWORD(buf, 1); // wType

        std::wstringstream strLangIdBuf;
        strLangIdBuf
            << std::uppercase
            << std::hex
            << std::setw(8)
            << std::setfill(L'0')
            << engLangId;
        const std::wstring strLangId = strLangIdBuf.str();
        write(buf, strLangId); // szKey
        add32bitPadding(buf); // Padding

        forEach(props, [&buf](const PropertyMap::value_type& entry) -> void {
            if (entry.first.rfind(L"FIXEDFILEINFO_", 0) == 0) {
                // Ignore properties to be used to initialize data in
                // VS_FIXEDFILEINFO structure.
                return;
            }

            // Fill String pseudo structure
            StreamSize stringLength(buf); // wLength
            writeWORD(buf, entry.second.size()); // wValueLength
            writeWORD(buf, 1); // wType
            write(buf, entry.first); // wKey
            add32bitPadding(buf); // Padding1
            write(buf, entry.second); // Value
            add32bitPadding(buf); // Padding2
        });
    } while (0);

    // Fill VarFileInfo pseudo structure
    StreamSize varFileInfoLength(buf); // wLength
    writeWORD(buf, 0); // wValueLength
    writeWORD(buf, 1); // wType
    write(buf, L"VarFileInfo"); // szKey
    add32bitPadding(buf); // Padding

    // Fill Var pseudo structure
    StreamSize varLength(buf); // wLength
    writeWORD(buf, sizeof(DWORD)); // wValueLength
    writeWORD(buf, 0); // wType
    write(buf, L"Translation"); // szKey
    add32bitPadding(buf); // Padding
    writeDWORD(buf, MAKELANGID(LANG_NEUTRAL, SUBLANG_NEUTRAL)); // Value
}


VS_FIXEDFILEINFO VersionInfo::createFIXEDFILEINFO() const {
    const ::FixedFileVersion fileVersion(props.find(
            L"FIXEDFILEINFO_FileVersion")->second);

    VS_FIXEDFILEINFO result;
    ZeroMemory(&result, sizeof(result));

    result.dwSignature = 0xFEEF04BD;
    result.dwStrucVersion = 0x00010000;
    result.dwFileOS = VOS_NT_WINDOWS32;
    result.dwFileType = VFT_APP;

    fileVersion.apply(result.dwFileVersionMS, result.dwFileVersionLS);

    PropertyMap::const_iterator entry = props.find(
            L"FIXEDFILEINFO_ProductVersion");
    if (entry == props.end()) {
        fileVersion.apply(result.dwProductVersionMS, result.dwProductVersionLS);
    } else {
        bool fatalError = false;
        try {
            const ::FixedFileVersion productVersion(entry->second);
            fatalError = true;
            productVersion.apply(result.dwProductVersionMS,
                    result.dwProductVersionLS);
        } catch (const std::exception&) {
            if (fatalError) {
                throw;
            }
            // Failed to parse product version as four component version string.
            fileVersion.apply(result.dwProductVersionMS,
                    result.dwProductVersionLS);
        }
    }

    return result;
}
